workflow/controller/operator.go

```diff
@@ -1802,7 +1802,7 @@ func (woc *wfOperationCtx) executeTemplate(ctx context.Context, nodeName string,
 	if retryNodeName != "" {
 		retryNode := woc.wf.GetNodeByName(retryNodeName)
 		if !retryNode.Fulfilled() && node.Fulfilled() { // if the retry child has completed we need to update outself
-			node, err = woc.executeTemplate(ctx, retryNodeName, orgTmpl, tmplCtx, args, opts)
+			retryNode, err = woc.executeTemplate(ctx, retryNodeName, orgTmpl, tmplCtx, args, opts)
```

Comment:
This part of the code is responsible for ensuring the retry node gets marked as completed. In the case of affected workflows in this issue, this retry node is the root node.

The code works correctly, however, it currently doesn't return the updated node (in this change `retryNode`), but the old node (`node`). Simply return the updated node so that code downstream can act on the completed root node instead of an old and incomplete reference.